README.md
## Anvil JSHint Plugin

This plugin requires anvil.js version 0.8.* or greater.

## Installation

```text
	anvil install anvil.jshint
```

## Usage

After you have installed the plugin you first need to reference the plugin inside the `dependencies` key of your `build.json`.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		// Required settings here...
	}
}
```

Then you will need to choose one of the following senarios in order for the plugin to know which files to lint.

* Linting all your files
* Linting Specifc Files
* Linting All Files Except

### Settings

#### Linting all your files

Let's say that you had 5 JavaScript files in your `src` folder and you wanted all of them to be linted when you build your project. The following `"all": true` setting will tell `anvil.jshint` that you want everything to be linted.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"all": true
	}
}
```

#### Linting Specific Files

If you had 5 JavaScript files in your `src` folder and you only wanted a specific subset of those to be linted, then you could use the `include` setting and provide a list of files you want to include in the linting process.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"include": [ "main.js" ]
	}
}
```

#### Linting All Files Except

If you had 5 JavaScript files in your `src` folder and you wanted most of them to be linted, but to ignore a couple of items, then you could use the `exclude` setting and provide a list of files you want to exclude in the linting process.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"exclude": [ "util.js" ]
	}
}
```

#### Breaking Build Option

By default if there are any errors that JSHint returns then the build process will be aborted. You can override this option by providing setting the `breakBuild` to `false`.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"all": true,
		"breakBuild": false
	}
}
```

#### Ignore Specific Errors

Sometimes there are JSHint errors that for one reason or another you want ignored by the build process. To do that you can provide an `ignore` option with a list of all the errors that you feel are acceptable for your project. You list the `line`, `character`, and `reason` (contains) of each error you'd like to ignore. You can provide a combination of these options for more or less flexibility.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"all": true,
		"ignore": [
			{ "file": "bad.js", line": 81, "character": 26, "reason": "'someVariable' is already defined." },
			... other options ...
		]
	}
}
```

The following option ignores reason for line 81 and character 26 in bad.js

```javascript
{ "file": "bad.js", line": 81, "character": 26, "reason": "'someVariable' is already defined." }
```

The following option ignores any error on line 81 and character 12 in bad.js

```javascript
{ "file": "bad.js", "line": 81, "character": 12 }
```

The following option ignores reason anywhere on line 81 in bad.js

```javascript
{ "file": "bad.js", "line": 81, "reason": "'someVariable' is already defined." }
```

The following option ignores any errors on line 81 in bad.js

```javascript
{ "file": "bad.js", "line": 81 }
```

The following option ignores any errors matching reason anywhere in bad.js

```javascript
{ "file": "bad.js", "reason": "literal notation" }
```

The following option ignores any errors matching reason anywhere in any file

```javascript
{ reason": "literal notation" }

#### JSHint Settings

You can always provide custom JSHint and global comments to the top of each of your JavaScript file to tweak it's lint settings, but that can be redundant and a nuisance. So, you can provide these common settings in your 'anvil.jshint' settings to be used during the linting process.

```javascript
{
	"source": "src",
	"spec": "spec",
	"output": [ "build" ],
	"dependencies" : [ "anvil.jshint" ],
	"anvil.jshint": {
		"all": true,
		"options": {
			"globals": {
				"console": false,
			},
			"white": true
		}
	}
}
```

## Uninstallation

```text
	anvil uninstall anvil.jshint
```

## TODOS

* .jshintrc in pwd then all way up, then $HOME
* Get unit test coverage once Anvil.js allows assertion of logs
* Make options/globals be on a per folder basis



changelog.md
## v0.0.6 - Oct 01, 2012

* Added custom error color support with the `colors` module
* Used `messageformat` module to add singluar/plural support in messages
* Add file property to the ignore options

spec/plugin.spec.js
var expect = require( "expect.js" ),
	_ = require( "underscore" ),
	api = require( "anvil.js" ),
	sinon = require( "sinon" );

describe( "When linting a good JavaScript file", function() {
	var messages = [], harness;

	before( function( done ) {
		harness = new api.PluginHarness( "anvil.jshint", "./" );
		console.log( harness.plugin );
		stub = sinon.stub( harness.plugin, "log", function( message ) {
			messages.push( message );
		});
		harness.addFile( "./src/test.js",
			'var food = "pizza";\n' +
			'if ( food === "pizza" ) {\n' +
			'  window.alert( "yeah!" );\n' +
			'}' );

		harness.addCommandArgs( "--jshint" );
		harness.buildOnly( function() {
			done();
		});
	});

	it( "should log a success message to the eventLog", function() {
		expect( messages ).to.have.length( 1 );
		expect( messages[ 0 ] ).to.be( "No issues Found." );
	});
});

src/plugin.js
var jshint = require( "jshint" ).JSHINT;
var colors = require( "colors" );
var MessageFormat = require( "messageformat" );

colors.setTheme({
	verbose: "cyan",
	info: "green",
	warn: "yellow",
	debug: "blue",
	error: "red",
	data: "grey"
});

var jshintFactory = function( _, anvil ) {
	_.str = require( "underscore.string" );
	_.mixin( _.str.exports() );

	return anvil.plugin({
		name: "anvil.jshint",
		activity: "pre-process",
		all: true, //false,
		inclusive: false,
		exclusive: false,
		breakBuild: true,
		ignore: [],
		fileList: [],
		commander: [
			[ "-hint, --jshint", "JSHint all JavaScript files" ]
		],
		settings: {},
		messageFormat: new MessageFormat( "en" ),

		configure: function( config, command, done ) {
			if ( !_.isEmpty( this.config ) ) {
				if ( this.config.all ) {
					this.all = true;
				} else if ( this.config.include && this.config.include.length  ) {
					this.inclusive = true;
					this.fileList = this.config.include;
				} else if ( this.config.exclude && this.config.exclude.length ) {
					this.exclusive = true;
					this.fileList = this.config.exclude;
				}
				this.settings = this.getSettings( this.config );
				this.breakBuild = this.config.breakBuild === false ?
					this.config.breakBuild : this.breakBuild;
				this.ignore = this.config.ignore || [];
			} else if ( command.jshint ) {
				this.all = true;
			}

			done();
		},

		getSettings: function( config ) {
			var settings = {};

			if ( config.options ) {
				settings.options = config.options;
				settings.globals = settings.options ? settings.options.globals : {};
				delete settings.options.globals;
			} else {
				// TODO: Look for .jshintrc options in pwd, up tree, or $HOME
			}

			return settings;
		},

		run: function( done ) {
			var jsFiles = [],
				options = {},
				that = this,
				totalErrors = 0,
				transforms, message;

			if ( this.inclusive ) {
				jsFiles = _.filter( anvil.project.files, this.anyFile( this.fileList ) );
			} else if ( this.all || this.exclusive ) {
				jsFiles = _.filter( anvil.project.files, function( file ) {
					return file.extension() === ".js";
				});
				if ( this.exclusive ) {
					jsFiles = _.reject( jsFiles, this.anyFile( this.fileList ) );
				}
			}

			if ( jsFiles.length > 0 ) {
				message = this.messageFormat.compile( "Linting {NUM_RESULTS, plural, one{one file} other{# files}}." );
				anvil.log.step( message({ "NUM_RESULTS": jsFiles.length }) );
				transforms = _.map( jsFiles, function( file ) {
					return function( done ) {
						that.lint( file, function( numberOfErrors ) {
							totalErrors += numberOfErrors;
							done();
						});
					};
				});
				anvil.scheduler.pipeline( undefined, transforms, function() {
					if ( that.breakBuild === true &&
						_.isNumber( totalErrors ) && totalErrors > 0 ) {
						message = that.messageFormat.compile( "project {NUM_RESULTS, plural, one{one error} other{# errors}}!" );
						anvil.events.raise( "build.stop", message({ "NUM_RESULTS": totalErrors }) );
					}
					done();
				});
			} else {
				done();
			}
		},

		anyFile: function( list ) {
			return function( file ) {
				return _.any( list, function( name ) {
					var alias = anvil.fs.buildPath( [ file.relativePath, file.name ] );
					return name === alias || ( "/" + name ) == alias;
				});
			};
		},

		lint: function( file, done ) {
			var that = this;

			anvil.log.event( "Linting '"+ file.fullPath + "'" );
			anvil.fs.read( [ file.fullPath ], function( content, err ) {
				if ( !err ) {
					that.lintContent( file, content, function( numberOfErrors ) {
						done( numberOfErrors );
					});
				} else {
					anvil.log.error( "Error reading " + file.fullPath + " for linting: \n" + err.stack  );
					done({});
				}
			});
		},

		lintContent: function( file, content, done ) {
			var result = jshint( content, this.settings.options || {}, this.settings.globals || {} ),
				validErrors = [], ignoredErrors = 0, that = this;

			if ( result ) {
				anvil.log.event( "No issues Found." );
				that.log( "No issues Found." );
			} else {
				validErrors = this.processErrors( file, jshint.errors, this.ignore || [] );
				_.each( validErrors, function( error ) {
					anvil.log.error( error );
					that.log( error );
				});
				message = this.messageFormat.compile( "{NUM_RESULTS, plural, one{# issue} other{# issues}} found." );
				anvil.log.event( message({ "NUM_RESULTS": validErrors.length }) );
				ignoredErrors = jshint.errors.length - validErrors.length;
				if ( this.ignore.length && ignoredErrors ) {
					message = this.messageFormat.compile( "{NUM_RESULTS, plural, one{# issue} other{# issues}} ignored." );
					anvil.log.event( message({ "NUM_RESULTS": ignoredErrors }) );
				}
			}

			done( validErrors.length );
		},

		processErrors: function( file, errors, ignore ) {
			var result = [], padding = {}, that = this;

			padding = {
				line: _.reduce( errors, function( memo, error ) {
					if ( error && error.line.toString().length > memo.toString().length ) {
						memo = error.line.toString();
					}
					return memo;
				}, "" ).length,
				character: _.reduce( errors, function( memo, error ) {
					if ( error && error.character.toString().length > memo.toString().length ) {
						memo = error.character.toString();
					}
					return memo;
				}, "" ).length
			};

			_.each( errors, function( error ) {
				if ( error ) {
					if ( error.evidence ) {
						if ( !that.isIgnorable( file, error, ignore ) ) {
							result.push( that.formatError( error, padding ) );
						}
					} else {
						result.push( "Too Many Errors!" );
					}
				}
			});

			return result;
		},

		formatError: function( error, padding ) {
			return "[L".data + _.lpad( error.line, padding.line, "0" ).data +
				":C".data + _.lpad( error.character, padding.character, "0" ).data +
				"] ".data + error.evidence.replace( /^\s*/g, "" ).italic.error + " -> ".data +
				error.reason.bold.error;
		},

		/*
		 * The following option ignores reason for line 81 and character 26 in bad.js
		 * { "file": "bad.js", line": 81, "character": 26, "reason": "'someVariable' is already defined." }
		 *
		 * The following option ignores any error on line 81 and character 12 in bad.js
		 * { "file": "bad.js", "line": 81, "character": 12 }
		 *
		 * The following option ignores reason anywhere on line 81 in bad.js
		 * { "file": "bad.js", "line": 81, "reason": "'someVariable' is already defined." }
		 *
		 * The following option ignores any errors on line 81 in bad.js
		 * { "file": "bad.js", "line": 81 }
		 *
		 * The following option ignores any errors matching reason anywhere in bad.js
		 * { "file": "bad.js", "reason": "literal notation" }
		 *
		 * The following option ignores any errors matching reason anywhere in any file
		 * { reason": "literal notation" }
		 */
		isIgnorable: function( file, error, ignoreList ) {
			var ignorable = false;

			ignorable = _.any( ignoreList, function( ignore ) {
				return file.fullPath.indexOf( ignore.file ) > -1 && error.line === ignore.line && error.character === ignore.character && error.reason.indexOf( ignore.reason ) > -1;
			});
			if ( !ignorable ) {
				ignorable = _.any( ignoreList, function( ignore ) {
					return file.fullPath.indexOf( ignore.file ) > -1 && error.line === ignore.line && error.character === ignore.character && !ignore.reason;
				});
			}
			if ( !ignorable ) {
				ignorable = _.any( ignoreList, function( ignore ) {
					return file.fullPath.indexOf( ignore.file ) > -1 && error.line === ignore.line && !ignore.character && error.reason.indexOf( ignore.reason ) > -1;
				});
			}
			if ( !ignorable ) {
				ignorable = _.any( ignoreList, function( ignore ) {
					return file.fullPath.indexOf( ignore.file ) > -1 && error.line === ignore.line && !ignore.character && !ignore.reason;
				});
			}
			if ( !ignorable ) {
				ignorable = _.any( ignoreList, function( ignore ) {
					return file.fullPath.indexOf( ignore.file ) > -1 && !ignore.line && !ignore.character && error.reason.indexOf( ignore.reason ) > -1;
				});
			}
			if ( !ignorable ) {
				ignorable = _.any( ignoreList, function( ignore ) {
					return !ignore.file && !ignore.line && !ignore.character && error.reason.indexOf( ignore.reason ) > -1;
				});
			}

			return ignorable;
		},

		log: function( message ) {
			console.log( message );
		}
	});
};

module.exports = jshintFactory;